Sprint 07

Fensiz opened the pull request:
Выполнена работа по спринту 7

Fensiz committed:
добавлено несколько тестов

Fensiz committed:
вынесена бизнес логика в presenter

Fensiz committed:
выполнена работа по 7 спринту

Fensiz committed:
мелкие правки

Andreynnt commented on ArrayTests/ArrayTests.swift:
```diff
@@ -0,0 +1,32 @@
+@testable import MovieQuiz
+
+final class ArrayTests: XCTestCase {
+	func testGetValueInRange() throws {
```
Тут и в других местах съезжают отступы, советую выделить этот код или даже весь файл и нажать: `ctrl`+`i`, и xcode все отформатирует, чтобы код соответствовал codestyle-у

Fensiz replied:
я использую ctrl+i, у меня так настроены отступы

Andreynnt commented on ArrayTests/ArrayTests.swift:
```diff
@@ -0,0 +1,32 @@
+		let array: [Int] = [1, 2, 3]
+
+		//When
+		let value = array[2]
```
Тут нужно протестировать array[safe: 2], нативный вариант смысла тестировать нет, это уже сделали разработчики)

Andreynnt commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,122 @@
+		if questionFactory != nil {
+			self.questionFactory = questionFactory
+		} else {
+			self.questionFactory = QuestionFactory(moviesLoader: MoviesLoader(), delegate: self)
```
Обычно такие if-ы не делают, а всегда передают фабрики и прочие сервисы одним из этих способов. Предпочтительнее вариант через инит, потому что тогда можно удобно подменять реализацию в тестах

Andreynnt commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,122 @@
+		}
+		self.viewController = viewController
+		self.statisticService = statisticService
+		self.questionFactory?.loadData()
```
В ините не принято делать загрузку данных, потому что зачастую инит происходит раньше, чем viewController появляется на экране. Лучше во viewDidLoad

Fensiz replied:
это же инит презентера, не понял связи,
технически этот инит вызывается как раз во viewDidLoad

Fensiz replied:
в данном случае мне нужно чтобы загрузка была разовой, при создании объекта и инициализатор мне показалось хорошо для этого подходит.
прошу подробнее расписать, почему так не стоит делать.

Fensiz replied:
и для этого придется сделать дополнительный метод в протоколе, который в дальнейшем не понадобится

Fensiz replied:
оставлю эту правку до следующих замечаний

Andreynnt commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,122 @@
+	private weak var viewController: MovieQuizViewProtocol?
+	private var correctAnswers = 0
+	private let questionsAmount = 10
+	private var statisticService: StatisticServiceProtocol?
```
Сервис и фабрику стоит сделать let и неопциональными. А для их протоколов указать, что они `QuestionFactoryProtocol: AnyObject` - только для классов

Andreynnt commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,6 +1,6 @@
 import UIKit
 
-final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
+final class MovieQuizViewController: UIViewController {//}, QuestionFactoryDelegate {
```
Давай удалим {//}

Andreynnt commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -141,51 +71,50 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
-			self.currentQuestionIndex = 0
-			self.correctAnswers = 0
+	func show(result: QuizResultsViewModel) {
+		DispatchQueue.main.async { [weak self, weak presenter] in
```
Необязательно делать [weak presenter]. Достаточно [weak self], тогда можно обращаться к презентеру вот так без утечки self?.presenter.doSmth()

Fensiz replied:
я в курсе, мне показалось так будет нагляднее, какие минусы у данного варианта?

Andreynnt commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	private var currentQuestionIndex = 0 {
+		didSet {
```
Обычно отступы делают = 4 пробела, советую также настроить в xcode: Preferences -> line wrapping 

<img width="847" alt="image" src="https://github.com/user-attachments/assets/5fc8c4b1-1cca-4637-bcca-0292404b1e61" />

Fensiz replied:
у меня так и задано и я всегда делаю Ctrl + I, так что у меня ничего не уезжает, просто так настроены отступы.
если вопрос по многострочному литералу, то как сделать чтобы Autoident не двигал его влево я не знаю

Fensiz committed:
удален комментарий по замечанию

Fensiz committed:
изменена версия таргетов для тестов

Fensiz committed:
исправлен тест расширения массива

Fensiz committed:
внесены исправления по замечаниям

опциональные переменные сервиса и фабрики изменены на константы
в фабрику добавлен метод setDelegate
протокол сервиса ограничен ссылочными типами
удалены лишние устройста для тестов

Fensiz committed:
финальные правки

для тестов оставил 2 таргета: unit и UI
все тесты перенес в них

Fensiz committed:
добавлен test plan
